src/VAC/View.cpp

```diff
@@ -1087,46 +1093,221 @@ void View::PMRReleaseEvent(int action, double x, double y)
         emit allViewsNeedToUpdate();
     }
-    else if(action==CIRCLE_ACTION)
+    else if(action == CIRCLE_ACTION)
     {
-      //global()->setSculptRadius(0);
-      lastMousePos_ = QPoint(mouse_Event_X_,mouse_Event_Y_);
+        vac_->endRectangleOfSelection();
+        lastMousePos_ = QPoint(mouse_Event_X_,mouse_Event_Y_);
 
-      // pos = viewToScene(x,y); <- by now: the identity
-      QPointF pos = QPointF(x,y);
-      double xScene = pos.rx();
-      double yScene = pos.ry();
+        double radiusH = shapeHeight / 2;
+        double radiusW = shapeWidth / 2;
 
-      double w = global()->settings().edgeWidth();
-      bool debug = false;
-      if(!debug)
-      {
-        if(mouse_isTablet_ &&  global()->useTabletPressure())
-          w *= 2 * mouse_tabletPressure_; // 2 so that a half-pressure would get the default width
-      }
-
-      double radius = sqrt((xScene - shapeStartX) * (xScene - shapeStartX) + (yScene - shapeStartY) * (yScene - shapeStartY)) / 2;
-      double centerX = (xScene + shapeStartX) / 2;
-      double centerY = (yScene + shapeStartY) / 2;
-
-      double startX = (xScene + shapeStartX) / 2 + radius;
-      double startY = (yScene + shapeStartY) / 2;
-      vac_->beginSketchEdge(startX, startY, w, interactiveTime());
-
-      for (double deg = 0; deg <= M_PI * 2; deg += 0.01)
-      {
-        double x = radius * cos(deg);
-        double y = radius * sin(deg);
-
-        double newX = centerX + x;
-        double newY = centerY + y;
-        vac_->continueSketchEdge(newX, newY, w);
-      }
-      vac_->endSketchEdge();
-
-      emit allViewsNeedToUpdatePicking();
-      updateHoveredObject(mouse_Event_X_, mouse_Event_Y_);
-      emit allViewsNeedToUpdate();
+        double centerX = (leftX + rightX) / 2;
+        double centerY = (topY + bottomY) / 2;
+
+        double startX = centerX + radiusW;
+        double startY = centerY;
+        vac_->beginSketchEdge(startX, startY, w, interactiveTime());
+
+        for (double deg = 0; deg < M_PI * 2; deg += 0.01)
+        {
+            double x = radiusW * cos(deg);
+            double y = radiusH * sin(deg);
+
+            double newX = centerX + x;
+            double newY = centerY + y;
+            vac_->continueSketchEdge(newX, newY, w);
+        }
+        vac_->endSketchEdge();
+
+        emit allViewsNeedToUpdatePicking();
+        updateHoveredObject(mouse_Event_X_, mouse_Event_Y_);
+        emit allViewsNeedToUpdate();
+    }
+    else if(action == TRIANGLE_ACTION)
+    {
+        vac_->endRectangleOfSelection();
+        lastMousePos_ = QPoint(mouse_Event_X_,mouse_Event_Y_);
+
+        vac_->beginSketchEdge(leftX, bottomY, w, interactiveTime());
+        vac_->continueSketchEdge(leftX + shapeWidth / 2, topY, w);
+        vac_->endSketchEdge();
+
+        vac_->beginSketchEdge(leftX + shapeWidth / 2, topY, w, interactiveTime());
+        vac_->continueSketchEdge(rightX, bottomY, w);
+        vac_->endSketchEdge();
```

Comment:
I feel like we could do a for loop here and combine triangle to octagon into a single piece of code.

Reply:
Maybe you mean to combine triangle and hexagon, and rectangle and octagon?
In general I plan to some improve algorithms for drawing polygons.
But it works more or less wellat now and I propose done this in the next PRs.